lib/auth/init.go

```diff
@@ -484,6 +483,49 @@ func Init(cfg InitConfig, opts ...ServerOption) (*Server, error) {
 	return asrv, nil
 }
 
+func initSetAuthPreference(asrv *Server, newAuthPref services.AuthPreference) error {
+	storedAuthPref, err := asrv.GetAuthPreference()
+	if err != nil {
+		if !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		storedAuthPref = nil
+	}
+	shouldReplace, err := shouldInitReplaceResource(storedAuthPref, newAuthPref)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if shouldReplace {
+		err = asrv.SetAuthPreference(newAuthPref)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		log.Infof("Updating auth preference: %v.", newAuthPref)
+	}
+	return nil
+}
+
+// shouldInitReplaceResource determines whether the candidate resource should
+// be used to replace the stored resource during auth server initialization.
+//
+// For more details, see Section "Auth server initialization"
+// in rfd/0016-dynamic-configuration.md
+func shouldInitReplaceResource(stored, candidate types.ResourceWithOrigin) (bool, error) {
+	if candidate.Origin() != types.OriginConfigFile && candidate.Origin() != types.OriginDefaults {
+		return false, trace.BadParameter("candidate origin must be either config-file or defaults (this is a bug)")
```

Comment:
Where does the `candidate` resource come from? If I'm reading correctly, it's coming from the configuration file? If so, users may see this error and it will be confusing to them.

Reply:
See above. This is just a sanity check for the resource-aware resources to come with properly set origin values. The error should be triggered only during development.